File: frontend/src/services/items.ts
import api from '@/lib/api'
import { Item, ItemAvailability } from '@/types'

export interface ItemFilters {
  search?: string
  category?: string
  subcategory?: string
  availability_type?: string
  neighborhood?: string
  city?: string
  lat?: number
  lng?: number
  lat2?: number
  lng2?: number
  radius_km?: number
  sort?: 'nearest' | 'price_asc'
  skip?: number
  limit?: number
}

export interface ItemPayload {
  title: string
  description?: string
  category: string
  subcategory?: string
  photos?: string[]
  availability_type: string
  daily_rate?: number | null
  weekly_rate?: number | null
  monthly_rate?: number | null
  delivery_fee?: number | null
  declared_value?: number | null
  usage_rules?: string
  zip_code?: string
  neighborhood?: string
  city?: string
  state?: string
  latitude?: number
  longitude?: number
  group_ids?: string[]
  is_public?: boolean
  available_days?: number[]
  requires_identity_verification?: boolean
  fulfillment_options?: ('pickup' | 'delivery')[]
  quantity_total?: number
}

export const itemsService = {
  list: (params?: ItemFilters) =>
    api.get<Item[]>('/items', { params }).then((r) => r.data),

  get: (id: string) => api.get<Item>(`/items/${id}`).then((r) => r.data),

  create: (data: ItemPayload) => api.post<Item>('/items', data).then((r) => r.data),

  update: (id: string, data: Partial<ItemPayload>) =>
    api.put<Item>(`/items/${id}`, data).then((r) => r.data),

  delete: (id: string) => api.delete(`/items/${id}`),

  uploadPhoto: (itemId: string, file: File) => {
    const formData = new FormData()
    formData.append('file', file)
    return api.post<Item>(`/items/${itemId}/photos`, formData).then((r) => r.data)
  },

  activate: (id: string) => api.patch<Item>(`/items/${id}/activate`).then((r) => r.data),

  deactivate: (id: string) => api.patch<Item>(`/items/${id}/deactivate`).then((r) => r.data),

  myItems: () => api.get<Item[]>('/users/me/items').then((r) => r.data),

  favorite: (id: string) => api.post<Item>(`/items/${id}/favorite`).then((r) => r.data),
  unfavorite: (id: string) => api.delete<Item>(`/items/${id}/favorite`).then((r) => r.data),
  myFavorites: () => api.get<Item[]>('/users/me/favorites').then((r) => r.data),

  joinWaitlist: (id: string) => api.post<Item>(`/items/${id}/waitlist`).then((r) => r.data),
  leaveWaitlist: (id: string) => api.delete<Item>(`/items/${id}/waitlist`).then((r) => r.data),

  checkAvailability: (id: string, pickupDate: string, expectedReturnDate: string) =>
    api
      .get<ItemAvailability>(`/items/${id}/availability`, {
        params: { pickup_date: pickupDate, expected_return_date: expectedReturnDate },
      })
      .then((r) => r.data),
}

File: web/app/routers/items.py
from fastapi import APIRouter, BackgroundTasks, Depends, File, Query, UploadFile

from app.dependencies import get_current_user, get_current_user_optional
from app.models.user import User
from app.schemas.item import ItemCreate, ItemResponse, ItemUpdate
from app.services import item_service

router = APIRouter(prefix="/items", tags=["items"])


@router.get("/", response_model=list[ItemResponse])
def list_items(
    search: str | None = Query(
        None, description="Full-text search over title and description"
    ),
    category: str | None = Query(None),
    subcategory: str | None = Query(None),
    availability_type: str | None = Query(None, pattern="^(free|paid)$"),
    neighborhood: str | None = Query(None),
    city: str | None = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(16, ge=1, le=100),
    lat: float | None = Query(None),
    lng: float | None = Query(None),
    radius_km: float | None = Query(None, ge=0.1),
    current_user: User | None = Depends(get_current_user_optional),
):
    """Browses the public item catalog with optional search, filters, and
    distance-based radius search."""
    return item_service.list_items(
        search=search,
        category=category,
        subcategory=subcategory,
        availability_type=availability_type,
        neighborhood=neighborhood,
        city=city,
        skip=skip,
        limit=limit,
        lat=lat,
        lng=lng,
        radius_km=radius_km,
        current_user=current_user,
    )


@router.get("/{item_id}", response_model=ItemResponse)
def get_item(
    item_id: str, current_user: User | None = Depends(get_current_user_optional)
):
    """A single item's public detail page."""
    return item_service.get_item(item_id, current_user)


@router.post("/", response_model=ItemResponse, status_code=201)
def create_item(data: ItemCreate, current_user: User = Depends(get_current_user)):
    """Lists a new item. Paid items require the owner to already have a
    Mercado Pago account connected."""
    return item_service.create_item(data, current_user)


@router.put("/{item_id}", response_model=ItemResponse)
def update_item(
    item_id: str, data: ItemUpdate, current_user: User = Depends(get_current_user)
):
    """Partial update of an item the logged-in user owns."""
    return item_service.update_item(item_id, data, current_user)


@router.delete("/{item_id}", status_code=204)
def delete_item(item_id: str, current_user: User = Depends(get_current_user)):
    """Deactivates (soft-deletes) an item the logged-in user owns."""
    item_service.delete_item(item_id, current_user)


@router.patch("/{item_id}/activate", response_model=ItemResponse)
def activate(
    item_id: str,
    background_tasks: BackgroundTasks,
    current_user: User = Depends(get_current_user),
):
    """Marks an item available again — clears its waitlist and notifies
    anyone who was on it."""
    return item_service.set_availability(item_id, True, current_user, background_tasks)


@router.patch("/{item_id}/deactivate", response_model=ItemResponse)
def deactivate(item_id: str, current_user: User = Depends(get_current_user)):
    """Marks an item temporarily unavailable, without deleting it."""
    return item_service.set_availability(item_id, False, current_user)


@router.post("/{item_id}/favorite", response_model=ItemResponse)
def favorite_item(item_id: str, current_user: User = Depends(get_current_user)):
    """Adds an item to the logged-in user's favorites."""
    return item_service.set_favorite(item_id, current_user, True)


@router.delete("/{item_id}/favorite", response_model=ItemResponse)
def unfavorite_item(item_id: str, current_user: User = Depends(get_current_user)):
    """Removes an item from the logged-in user's favorites."""
    return item_service.set_favorite(item_id, current_user, False)


@router.post("/{item_id}/waitlist", response_model=ItemResponse)
def join_waitlist(item_id: str, current_user: User = Depends(get_current_user)):
    """Joins the wait list for an unavailable item — notified once it's
    marked available again."""
    return item_service.set_waitlist(item_id, current_user, True)


@router.delete("/{item_id}/waitlist", response_model=ItemResponse)
def leave_waitlist(item_id: str, current_user: User = Depends(get_current_user)):
    """Leaves the wait list for an item."""
    return item_service.set_waitlist(item_id, current_user, False)


@router.post("/{item_id}/photos", response_model=ItemResponse, status_code=201)
async def upload_photo(
    item_id: str,
    file: UploadFile = File(...),
    current_user: User = Depends(get_current_user),
):
    """Uploads a photo for an item the logged-in user owns — resized and
    stripped of EXIF/GPS metadata before storage."""
    return await item_service.upload_photo(item_id, file, current_user)
